### DISCUSSION TITLE
Acceptance testing provider using OpenTofu CLI

### DISCUSSION THREAD

**Discussion by alexhung** [opentofu/opentofu#975] [Q&A] [answered]
Hi, I maintain a few providers for my employer and we are investing how to support OpenTofu in our development process. One main issue we come across is executing acceptance tests using OpenTofu CLI.

I set the env var `TF_ACC_TERRAFORM_PATH` per [documentation](https://github.com/orgs/opentofu/discussions/975) with the path to the OpenTofu CLI (`/usr/local/bin/tofu`). When I run the acceptance tests with no change to our code, I got the following error:

```
=== RUN   TestAccRemoteUpgradeFromVersionWithNoDisableProxyAttr
    resource_artifactory_remote_repository_test.go:42: TestStep 2/2 running init: exit status 1
        Error parsing %!q(PANIC=String method: called String on zero-value addrs.Provider) as a provider address: Invalid provider namespace: The legacy provider namespace "-" can be used only with hostname registry.opentofu.org.
    testing_new.go:79: Error retrieving state, there may be dangling resources: exit status 1
        Error parsing %!q(PANIC=String method: called String on zero-value addrs.Provider) as a provider address: Invalid provider namespace: The legacy provider namespace "-" can be used only with hostname registry.opentofu.org.
```

Is there extra steps I need to do or code to change in order to test against OpenTofu CLI?

Would there be a migration doc/guide for provider developers like for regular users?

**Comment by ghost** [comment]
Hey @alexhung this is really useful, thank you. I [opened an issue](https://github.com/opentofu/opentofu/issues/977) from this discussion as it seems to be a bug. We were debating if the legacy namespace thing was still in use or if it's something long gone. This should work out of the box and we'll look into why it's broken. /cc @Yantrio @cube2222 

To answer your question about the provider development flow, ideally it should be the same process, but if it's more than just pointing `TF_ACC_TERRAFORM_PATH` to tofu, then that would indeed be something we should do.

**Comment by Yantrio** [comment, marked as answer]
Hi @alexhung thank you for raising this.

I've tracked this down to a choice of hardcoding some parameters in the plugin testing code [here](https://github.com/hashicorp/terraform-plugin-testing/blob/76a36e49f4ccb8043f8641455ffdb104dd95e4cb/helper/resource/plugin.go#L153-L162)

To work around this, when testing with tofu you can set the following environment variables:
```
TF_ACC_TERRAFORM_PATH="/path/to/opentofu"
TF_ACC_PROVIDER_NAMESPACE="hashicorp"
TF_ACC_PROVIDER_HOST="registry.opentofu.org"
```

I have successfully cloned and run the `TestAccRemoteUpgradeFromVersionWithNoDisableProxyAttr` test from your provider repository using these environment variables. 

Please try this out and let us know how you get on! I'm happy to help with any other snags you hit whilst doing acceptance testing of providers with OpenTofu.

**Reply by alexhung** [reply]
@Yantrio I can confirm that with those 3 env vars set, the entire acceptance test suite is executed successfully!

Please add this to documentation for provider developers 😄

**Reply by ghost** [reply]
Raised an issue for the docs [here](https://github.com/opentofu/opentofu/issues/982).
